### DISCUSSION TITLE
Issue with 32-bit devices

### DISCUSSION THREAD

**Discussion by sanchitgarg1909** [krzyzanowskim/CryptoSwift#961] [General]
In my latest update, in 32-bit devices the encrypted data being generated is corrupted. When I try to decrypt the encrypted data it fails. So the users are not able to use the app at all. I didn't change anything on my encrypt or decrypt logic. It is suddenly happening in the latest release. I'm currently using Crypto version 0.9.0. Anybody else facing similar issue?

**Comment by sidorov-panda** [comment]
Hey @sanchitgarg1909 , seems like I have a similar issue. I'm using 0.15.0.
Trying to figure out how to deal with it. Please, update me if you find a solution

**Comment by sanchitgarg1909** [comment]
Hi @sidorov-panda. These are my current findings :-
1. It is affecting only 32-bit devices iPad 4, iPhone 5/5c in my case.
2. Apple released a silent update 10.3.4 last month. This could be causing the issue. 
I'm currently using Crypto 0.9.0. Which version are you using? When did this issue start happening for you? Are you able to reproduce this issue?

**Comment by sidorov-panda** [comment]
I'm using 0.15.0, I found the issue yesterday, it seems like it affects only new builds, I tried old versions from TestFlight and it works just fine. I can reproduce it only on an old iPhone 5, other devices seems to be good

**Comment by sanchitgarg1909** [comment]
I don't have an old iPhone 5 or iPad 4 so I haven't been able to reproduce the issue. I also asked one of the affected users to download older version from testflight and it worked fine. I haven't changed the encryption logic or my Crypto version in these 2 app versions so I'm confused how is this happening. My plan is to create a build with the latest Crypto 1.0.0 and check if that works. I'll update you.

**Comment by sanchitgarg1909** [comment]
Update - Updating to the latest version of Crypto did not solve the issue. The only solution right now is to distribute the previous version of the app to the affected users. If this issue does not get fixed, we will stop supporting 32 bit devices in future updates.

**Comment by yarodevuci** [comment]
Same issue happened to me. I am on 1.0.0 with new build. So should I roll back to old version ?

**Comment by sanchitgarg1909** [comment]
@yarodevuci You can try it but I don't think it will work.

**Comment by yarodevuci** [comment]
@sanchitgarg1909 users on iphone 8 so far I noticed have issues. I wasn't able to reproduce it on simulator. That's strange

**Comment by sanchitgarg1909** [comment]
@yarodevuci For me the issue is only happening in 32 bit devices. But it isn't reproducible on simulator. Can we get in touch with one of CryptoSwift devs?

**Comment by yarodevuci** [comment]
> @yarodevuci For me the issue is only happening in 32 bit devices. But it isn't reproducible on simulator. Can we get in touch with one of CryptoSwift devs?

Most likely because simulators are run on 64 bit computers. That's why it's not reproducible. 

@krzyzanowskim can you help us out?

Apple released a silent update 10.3.4 last month. What is that update?? @sanchitgarg1909

**Comment by sanchitgarg1909** [comment]
@yarodevuci It is a security update. My initial assumption was also that the OS update might have caused the issue. But it is also observed on devices running 10.3.3

**Comment by objectivecosta** [comment]
Hello, everyone. I'm experiencing a bug with iPad 2 ATM, where a few bits get changed when calling -encrypted, resulting in a corrupted encrypted data. Example:

iOS 9; iPad 2; 32-bit device:        `8DE3ED1488FD7F5BA451F21E28D8CA71`
iOS 13; iPhone X; 64-bit device: `8DF73B1488F5565BA47D8F1E28D8D771`

Decryption works fine, when the input is correct. Does this match the bug behaviour you've been seeing? At first I thought it was related to iOS 9, but it might be this one.

Best,

Rafael

**Comment by NutdanaiVankrua** [comment]
Having same issue with 32 bit devices (iPhone 4/ 4s/ 5/ 5c) But when run the same code on 64 bit devices no issues at all.

**Comment by objectivecosta** [comment]
Related: https://forums.developer.apple.com/thread/123044

**Comment by sanchitgarg1909** [comment]
I'm using Xcode 10.3, is anyone facing the issue in the same? I'm not sure if this issue started in Xcode 10.3 and was fine in Xcode 10.2.1. Since I don't have a 32 device to check, can anyone confirm this? Would be a lot of help.

**Comment by yarodevuci** [comment]
@sanchitgarg1909 Mine was compiled on XCode 10.3 as well. I should try recompile on 10.2.1 and push it on production to see. I don't have 32bit device either.

**Comment by sanchitgarg1909** [comment]
@yarodevuci Do you remember if your previous app version which was working fine was compiled in which Xcode version? 
I'm uploading a build from Xcode 10.2.1 and giving it to an affected user using Testflight. I'll let you know if it works.

**Comment by yarodevuci** [comment]
@sanchitgarg1909 
Xcode 10.3 was released on July 22nd, last working update for my app was released a month ago. It most likely was compiled on 10.3 XCode version. I updated the CryptoSwift to latest version before next release. But previously was said that previous version also have issues. So am not sure. I feel like users on iOS13 only have issues.

**Comment by krzyzanowskim** [comment]
Can anybody check if this unexpected solution helps in your case as well? https://github.com/krzyzanowskim/CryptoSwift/issues/752#issuecomment-535604917

**Comment by objectivecosta** [comment]
@krzyzanowskim I can reproduce the issue both at compile time locally (DEBUG) and production. I don't think that it is the same issue.

**Comment by krzyzanowskim** [comment]
@rafaelc0sta can you maybe link here a test project that reproduces the issue always? and on what device/simulator

**Comment by nicolasgomollon** [comment]
@krzyzanowskim I just wanted to chime in here. We’re facing the same issue with our users running 32-bit devices recently. Luckily, one of those users is in our TestFlight group, and he’s been responsive, so we’ve pushed out two identical builds to him, one with _Bitcode enabled_ at upload time, and one with _Bitcode disabled_ at upload time, and he only experienced the issue with the build that was uploaded with _Bitcode enabled_. So, essentially the https://github.com/krzyzanowskim/CryptoSwift/issues/752#issuecomment-535604917 is able to solve this issue for now. We’re currently using Xcode 10.3, but I don’t think it matters, since Apple can recompile from the Bitcode we submit at any time.

**Comment by nicolasgomollon** [comment]
@krzyzanowskim As for reproducing the issue, you’ll need a 32-bit device _physically_ (the simulator will not reproduce the issue), and submitting a build to either TestFlight or App Store with Bitcode enabled at upload time.

**Comment by krzyzanowskim** [comment]
@nicolasgomollon Bitcode issue is likely due to bitcode miscompilation bug that happens at AppStore side. The problem is reported to Apple. I tried to escalate however I can, but I don't know if anyone is working on it actively or what is the solution here.

However, @@rafaelc0sta suggest that there's a different issue, not related to bitcode. I'd like to investigate and reproduce this one.

**Comment by krzyzanowskim** [comment]
@nicolasgomollon would you mind reporting that to Apple via https://feedbackassistant.apple.com/ and link this thread in the description?

**Comment by sanchitgarg1909** [comment]
Can anyone help me out with the following :-
1. Is disabling Bitcode a recommended way? 
2. Can it affect my app in any other way?
3. Is it better to not use this solution and just wait for Apple to fix it?

**Comment by krzyzanowskim** [comment]
@sanchitgarg1909 disable bitcode. report to apple. link this thread in bugreport. hopefully somebody from apple will reach to you or me to helpwith the issue

**Comment by krzyzanowskim** [comment]
Did anyone get any feedback from Apple on the report?

**Comment by sanchitgarg1909** [comment]
Hi all,
The workaround of disabling the bitcode worked for me earlier. However, my today's release again had the same issue even after disabling the bitcode. Has the workaround also stopped working? I used Xcode 11 to create the archive, maybe the same bug is in Xcode 11.

**Comment by krzyzanowskim** [comment]
That's bad. No, darn idea. I've got no update. Let's try to narrow down the code. What code (or classess) do you use?

**Comment by jckarter** [comment]
By 32-bit devices, do you mean armv7k Apple Watches, or both armv7k and arm64_32 watches?

**Comment by nicolasgomollon** [comment]
@jckarter Not sure about the watches, but I can confirm it breaks on iPhone 5c (armv7)

**Comment by sanchitgarg1909** [comment]
Update - Finally able to solve this issue by creating a build from Xcode 10 and unchecking the "include bitcode" option. All developers please do not release any app update from Xcode 11.
@krzyzanowskim - I use sha256 as the hash function. For encryption I use AES with CBC as blockmode and pkcs7 as padding.

**Comment by objectivecosta** [comment]
I ended up rewriting all the crypto in my app to use CommonCrypto. It was a pain, but sadly I have to support 32-bit devices. ¯\_(ツ)_/¯ 

I use AES, with CBC as block mode and also PKCS#7 padding, just like @sanchitgarg1909

**Comment by krzyzanowskim** [comment]
@rafaelc0sta if you can reproduce it, can you give a try current `master` branch maybe?

**Comment by krzyzanowskim** [comment]
I got informed that the LLVM team found the issue and applied patch, and that it will be pushed to Xcode 11 as well.

Fix: https://github.com/llvm/llvm-project/commit/dd8cd6d26b1706d83d4aea0c09e46f0e21195de5

**Comment by nicolasgomollon** [comment]
@krzyzanowskim Do you have any details as to whether we will have to recompile our project to include this fix, or will Apple retroactively recompile the affected builds on their end with the supplied Bitcode?

**Comment by krzyzanowskim** [comment]
I don't think Apple will recompile anything. However, once they update internal tools, and if in fact that's something that was caused by bitcode transpilation, then new uploads should fix automatically. Maybe @jckarter know more what are the chances here.

**Comment by jckarter** [comment]
I don't know when the bitcode build tool gets revved exactly, but when it does, then existing bitcode submissions should get rebuilt with the fix automatically.

**Comment by thomasnordquist** [comment]
Updating XCode to 11.2.1 did not fix the issue for us.
Updating from `1.0.0` to `1.1.3` fixed our issues. (`1.1.1` also worked)

Device: iPhone 5 (A1429), iOS 10.3.4

**Comment by krzyzanowskim** [comment]
That's great news!

Can somebody else confirm here maybe, before I close it?

**Comment by paulocesar-ufu** [comment]
Yes, everything ok here:
Xcode 11
CryptoSwift 1.1.3

No problem encrypting in 32bit devices.

**Comment by krzyzanowskim** [comment]
Thank you! Closing.

**Comment by nishanair23791** [comment]
Getting error as module compiled with 4.0.3 cannot be imported to swift 5.1 compiler while archive . Kindly help me out with this

**Comment by objectivecosta** [comment]
> Getting error as module compiled with 4.0.3 cannot be imported to swift 5.1 compiler while archive . Kindly help me out with this

@nishanair23791 This is a thread for another issue. Your issue is not related to CryptoSwift and it is not the same issue as the one in the thread. I'd recommend you open an issue on the appropriate channels. 

That said, you're compiling modules with one compiler version and importing it into another. You shouldn't do that. Change everything to the same Swift compiler version and you should be fine.

**Comment by nishanair23791** [comment]
Actually I m stuck up with the same issue of 32 bit device since 1 month . And I have taken code of cryptoswift with 4.2 version and build is getting succeeded but not able to archive

**Comment by objectivecosta** [comment]
@nishanair23791 Update your Xcode version and CryptoSwift version. The bug should be fixed now, as said in the thread. I too confirmed it myself on my apps.

**Comment by nishanair23791** [comment]
I’m using Xcode 11 is there any issue with that?

**Comment by nishanair23791** [comment]
Do I need to update my code to swift 5.1 for new update of crypto swift ,will it not work on swift 4.2 . Kindly confirm.

**Comment by krzyzanowskim** [comment]
@nishanair23791 yes. Check README.md for information about swift version vs cryptoswift version

**Comment by nishanair23791** [comment]
Issue in 32 bit device with App Store distribution profile is still there , Adhoc and development profile it is working fine . I m using crypto swift 4.2 version and Xcode 11 . Is there any work around to fix this issue for crypto swift 4.2 branch update .

**Comment by krzyzanowskim** [comment]
> Issue in 32 bit device with App Store distribution profile is still there

it's not

> I m using crypto swift 4.2 version and Xcode 11 .

you should update to CryptoSwift 1.2.x or newer as it was stated before https://github.com/krzyzanowskim/CryptoSwift/issues/748#issuecomment-559684746
